modules/gapi/include/opencv2/gapi/rmat.hpp

```diff
@@ -0,0 +1,129 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2020 Intel Corporation
+
+#ifndef OPENCV_GAPI_RMAT_HPP
+#define OPENCV_GAPI_RMAT_HPP
+
+#include <opencv2/gapi/gmat.hpp>
+
+namespace cv {
+
+// "Remote Mat", a general class which provides an abstraction layer over the data
+// storage and placement (host, remote device etc) and allows to access this data.
+//
+// The device specific implementation is hidden in the RMat::Adapter class
+//
+// The basic flow is the following:
+// * Backend which is aware of the remote device:
+//   - Implements own AdapterT class which is derived from RMat::Adapter
+//   - Wraps device memory into RMat via make_rmat utility function:
+//         cv::RMat rmat = cv::make_rmat<AdapterT>(args);
+//
+// * End user:
+//   - Writes the code which works with RMats without any knowledge of the remote device:
+//     void func(const cv::RMat& in_rmat, cv::RMat& out_rmat) {
+//         // Fetch input data from the device, get mapped memory for output
+//         cv::RMat::View  in_view =  in_rmat.access(Access::R);
+//         cv::RMat::View out_view = out_rmat.access(Access::W);
+//         performCalculations(in_view, out_view);
+//         // data from out_view is transferred to the device when out_view is destroyed
+//     }
+class RMat
+{
+public:
+    // A lightweight wrapper on image data:
+    // - Doesn't own the memory;
+    // - Doesn't implement copy semantics (it's assumed that a view is created each time
+    // wrapped data is being accessed);
+    // - Has an optional callback which is called when the view is destroyed.
```

Comment:
Would argue that specifically `std::unique_ptr` is a good idea, it's another level of indirection that is questionable here (+ a guaranteed heap allocation, not that it matters performance-wise given the nature of `RMat` but still).
Also it looks to me that `View` is mostly about point 2. here. The mimicking of cv::Mat APIs is done for convenience rather than for anything else. Yet, a very fair point.

To me this whole thing looks very similar to a pattern from generic locking. Applicable to the domain of view, my proposal is the following:

```cpp
class View {
  cv::Mat m_non_owning_mat;  // this is a cv::Mat created on external pointer, the external pointer is returned by "adapter"/RMat
  CallbackType m_cb;

  View(GMatDesc desc, uchar* data, size_t step, ...) : m_non_owning_mat(desc, step, data, ...) {}  // View's ctor constructs cv::Mat from given parameters

  ~View() {  if (m_cb) m_cb();  }  // calls callback on destruction

   cv::Mat* operator->() { return &m_non_owning_mat; }  // the idea of this becomes clear from the usage model

   cv::Mat& operator*()  { return m_non_owning_mat; }  // I think you might need this as well in certain contexts
};

// usage:
auto view = create_view();
view->elemSize();  // returns the value, returned by cv::Mat::elemSize();
view->depth();  // returns the value, returned by cv::Mat::depth();
view->at(pos);  // returns the pixel/row, returned by cv::Mat::at(pos);
```

The idea for `operator->()` is based around the fact that for a given expression `x->something`, the compiler would unwrap a call to `->` into a sequence of calls to `operator->()` until it happens to find a raw pointer. Thus, in our case
`view->elemSize()` in an unwrapped form would look something like `(view->operator())->elemSize()`. This is a nicely abusable property used in many different domains (locking patterns, logging, introspection, smart pointers use that as well, proxy classes?, etc.).

@anton-potapov  would the proposed design look good in terms of single responsibility, what do you think?
@rgarnov please also share your opinion.

By the way, having `operator->()` trick would eliminate the need to mimic any cv::Mat API as we could use cv::Mat directly.

Reply:
Also you can go an extra mile and figure a nice way to provide const-method-only view for `Access::R` and non-const view for `Access::W`

Reply:
> Mimics the cv::Mat 

NO

Reply:
> (e.g. to std::unique_ptr with a custom deleter)

..so then a user could do `ptr.release()` and say good-bye to this custom deleter. Let's avoid this leg-shooting at an early stage.